### ISSUE TITLE
[check-font] define check grouping schema and apply to all checks

### ISSUE THREAD

**Issue by m4rc1e** [fonttools/fontbakery#1574]
We currently use a single number to identify tests. ID numbers are assigned to checks as soon as they are created. Unfortunately this approach cannot handling grouping tests.

Our current approach work as follows:
```
1. Check fontname          (added on day 1)
2. Check missing glyphs    (added on day 2)
3. Check style name        (added on day 3)
4. Check missing styles    (added on day 4)
```

I would prefer to see a system like this:
```
M1. Check fontname         (added on day 1)
M2. Check style name       (added on day 3)
R1. Check missing glyphs   (added on day 2)
R2. Check missing styles   (added on day 4)
```
By prepending a suffix at the beginning, it gives us more flexibility to group tests. In the above example "M" could be a suffix for metadata and "R" for regression check groups.

This is by no means the best solution. I just thought it would be a good idea to start a discussion for this. We currently have 120+ tests in a 2,500+ python file. We need to start thinking about how to structure this better and a decent id system will play a crucial role in this.

**Comment by felipesanches** [comment]
yes, I agree with that. I also dislike the purely numerical IDs.

Maybe a word prefix would be better than a letter prefix. For instance, for tests that are specific to certain tables, the table name could be used as a prefix. So, for name table tests we could have **com.gogle.fonts/test/name/1**, **com.gogle.fonts/test/name/2**, **com.gogle.fonts/test/name/3**, etc...

A neat thing then would be the ability to use "-c com.gogle.fonts/test/name" to select a whole group of tests.

If we ever change the test-id naming scheme, we should make sure of keeping a reference to the old id as well in order to make it easier to track down project history on the future.

**Comment by felipesanches** [comment]
I also dislike having a whole bunch of third-party checks (such as fontforge, FontVal, etc) being logged as results of single FB tests. It would be neat to map those to individual FB check results.

But this perhaps would be much more challenging to implement given the nature of our current third-party tools integration strategy.

**Comment by graphicore** [comment]
Grouping tests and ids should be something different. IDs are rather meaningless, they just need to be unique, forever. Don't overload IDs. Think of IDs like this: I want to create a URL that is valid in 10 years ... could be a random string, could be the date of today plus a word, could be a [Reverse domain name notation](https://en.wikipedia.org/wiki/Reverse_domain_name_notation) (which has some upsides).

Fringe: If we manage to create a globally agreed upon collection of tests, and therefore need to move test somewhere else, ideally we wouldn't change the ID as long as it is semantically the [same …](https://www.youtube.com/watch?v=vAUOPHqx5Gs)

> If we ever change the test-id naming scheme, we should make sure of keeping a reference to the old id as well in order to make it easier to track down project history on the future.

We have one shot at this, now, and we'd already loose git history coherence. Generally:

Just don't change the old IDs **EVER**! We can come up with grouping mechanisms, like "tags" or literally "groups", we can come up with "nice" "human readable names" they can co-exist with stupid old IDs ;-) We can come up with new ID-schemes as long as they don't collide with the old ones.

We can NOW change these IDs, but as soon as we seriously go into production, and want to have a way to track a font-family history for 10 years into the past, these IDs should never change. Thus, don't overload IDs, make sure they are unique forever :-)


> I also dislike having a whole bunch of third-party checks (such as fontforge, FontVal, etc) being logged as results of single FB tests. It would be neat to map those to individual FB check results.
> 
> But this perhaps would be much more challenging to implement given the nature of our current third-party tools integration strategy.

No, it's easy! Run them as a "`condition`" and let the "`tests`" just yield the results you are interested in. Don't be scared away of the names these things have. A condition can run FontVal and return all results and several tests can cherry-pick from this data and just **yield** what they think is interesting.

**Comment by felipesanches** [comment]
yeah... we already do something like that with fontforge test results and I was considering doing more of that with other third-party tools as well.

**Comment by davelab6** [comment]
> Grouping tests and ids should be something different ... We can come up with grouping mechanisms, like "tags" or literally "groups", we can come up with "nice" "human readable names" they can co-exist with stupid old IDs ;-) 

I agree.

**Comment by felipesanches** [comment]
@m4rc1e any new thoughts about this issue nowadays ?

**Comment by felipesanches** [comment]
@m4rc1e ?

**Comment by simoncozens** [comment]
This is way out of date.
